File: attractor/pipeline/status.py
"""Status file management and run directory structure."""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any

from pydantic import BaseModel, Field


class NodeStatus(BaseModel):
    """Status of a completed node."""
    outcome: str = "success"  # success, retry, fail, partial_success
    preferred_next_label: str = ""
    suggested_next_ids: list[str] = Field(default_factory=list)
    context_updates: dict[str, Any] = Field(default_factory=dict)
    notes: str = ""


def write_status(node_id: str, status: NodeStatus, run_dir: Path) -> None:
    """Write status.json for a node."""
    node_dir = run_dir / node_id
    node_dir.mkdir(parents=True, exist_ok=True)
    target = node_dir / "status.json"
    target.write_text(status.model_dump_json(indent=2), encoding="utf-8")


def read_status(node_id: str, run_dir: Path) -> NodeStatus | None:
    """Read status.json for a node."""
    target = run_dir / node_id / "status.json"
    if not target.exists():
        return None
    try:
        data = json.loads(target.read_text(encoding="utf-8"))
        return NodeStatus(**data)
    except Exception:
        return None


def write_prompt(node_id: str, prompt: str, run_dir: Path) -> None:
    """Write prompt.md for a node."""
    node_dir = run_dir / node_id
    node_dir.mkdir(parents=True, exist_ok=True)
    (node_dir / "prompt.md").write_text(prompt, encoding="utf-8")


def write_response(node_id: str, response: str, run_dir: Path) -> None:
    """Write response.md for a node."""
    node_dir = run_dir / node_id
    node_dir.mkdir(parents=True, exist_ok=True)
    (node_dir / "response.md").write_text(response, encoding="utf-8")


def ensure_run_dir(run_dir: Path) -> None:
    """Ensure the run directory structure exists."""
    run_dir.mkdir(parents=True, exist_ok=True)
    (run_dir / "artifacts").mkdir(exist_ok=True)

File: internal/pipeline/status.go
package pipeline

import (
	"encoding/json"
	"os"
	"path/filepath"

	"github.com/nigelpepper/attractor/internal/llm"
)

// NodeStatus is the persisted result of a completed node.
type NodeStatus struct {
	Outcome            string         `json:"outcome"`
	PreferredNextLabel string         `json:"preferred_next_label"`
	SuggestedNextIDs   []string       `json:"suggested_next_ids"`
	ContextUpdates     map[string]any `json:"context_updates"`
	Notes              string         `json:"notes"`
	Usage              *llm.Usage     `json:"usage,omitempty"`
}

// WriteStatus writes status.json for a node.
func WriteStatus(nodeID string, status NodeStatus, runDir string) error {
	if status.Outcome == "" {
		status.Outcome = "success"
	}
	if status.SuggestedNextIDs == nil {
		status.SuggestedNextIDs = []string{}
	}
	if status.ContextUpdates == nil {
		status.ContextUpdates = map[string]any{}
	}
	dir := filepath.Join(runDir, nodeID)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	data, err := json.MarshalIndent(status, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(dir, "status.json"), data, 0o644)
}

// ReadStatus reads status.json for a node, or (nil, nil) if absent.
func ReadStatus(nodeID, runDir string) (*NodeStatus, error) {
	data, err := os.ReadFile(filepath.Join(runDir, nodeID, "status.json"))
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var s NodeStatus
	if err := json.Unmarshal(data, &s); err != nil {
		return nil, nil
	}
	return &s, nil
}

// WritePrompt writes prompt.md for a node.
func WritePrompt(nodeID, prompt, runDir string) error {
	dir := filepath.Join(runDir, nodeID)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(dir, "prompt.md"), []byte(prompt), 0o644)
}

// WriteResponse writes response.md for a node.
func WriteResponse(nodeID, response, runDir string) error {
	dir := filepath.Join(runDir, nodeID)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(dir, "response.md"), []byte(response), 0o644)
}

// EnsureRunDir creates the run directory structure.
func EnsureRunDir(runDir string) error {
	if err := os.MkdirAll(runDir, 0o755); err != nil {
		return err
	}
	return os.MkdirAll(filepath.Join(runDir, "artifacts"), 0o755)
}
